vision/mnist/DCGAN/dcgan.jl

```diff
@@ -0,0 +1,157 @@
+# Get the imports done
+using Flux, Flux.Data.MNIST
+using Flux: @epochs, back!, testmode!, throttle
+using Base.Iterators: partition
+using Distributions: Uniform,Normal
+using CUDAnative: tanh, log, exp
+using CuArrays
+using Images
+using Statistics
+
+# Define the hyperparameters
+BATCH_SIZE = 128
+NUM_EPOCHS = 15
+noise_dim = 100
+channels = 128
+hidden_dim = 7 * 7 * channels
+training_steps = 0
+verbose_freq = 100
+dis_lr = 0.0001f0 # Discriminator Learning Rate
+gen_lr = 0.0001f0 # Generator Learning Rate
+
+# Loading Data
+
+# We use Flux's built in MNIST Loader
+imgs = MNIST.images()
+
+# Partition into batches of size 'BATCH_SIZE'
+data = [reshape(float(hcat(vec.(imgs)...)),28,28,1,:) for imgs in partition(imgs, BATCH_SIZE)]
+
+# Define out distribution for random sampling for the generator to sample noise from
+dist = Normal(0.0,1.0) # Standard Normal noise is found to give better results
+
+expand_dims(x,n::Int) = reshape(x,ones(Int64,n)...,size(x)...)
+squeeze(x) = dropdims(x, dims = tuple(findall(size(x) .== 1)...))
+
+# The Generator
+generator = Chain(
+            Dense(noise_dim, 1024, leakyrelu), 
+            x->expand_dims(x,1),
+            BatchNorm(1024),
+            x->squeeze(x),
+            Dense(1024, hidden_dim, leakyrelu), 
+            x->expand_dims(x,1),
+            BatchNorm(hidden_dim),
+            x->squeeze(x),
+            x->reshape(x,7,7,channels,:),
+            ConvTranspose((4,4), channels=>64, relu; stride=(2,2), pad=(1,1)), 
+            x->expand_dims(x,2),
+            BatchNorm(64),
+            x->squeeze(x),
+            ConvTranspose((4,4), 64=>1, tanh; stride=(2,2), pad=(1,1))
+            ) |> gpu
+
+# The Discriminator
+discriminator = Chain(
+                Conv((3,3), 1=>32, leakyrelu;pad = 1), 
+                x->meanpool(x, (2,2)),
+                Conv((3,3), 32=>64, leakyrelu;pad = 1), 
+                x->meanpool(x, (2,2)),
+                x->reshape(x,7*7*64,:),
+                Dense(7*7*64, 1024, leakyrelu), 
+                x->expand_dims(x,1),
+                BatchNorm(1024),
+                x->squeeze(x),
+                Dense(1024, 1,sigmoid)
+                ) |> gpu
+
+# <b>Define the optimizers</b>
+
+opt_gen  = ADAM(params(generator),gen_lr, β1 = 0.5)
+opt_disc = ADAM(params(discriminator),dis_lr, β1 = 0.5)
+
+# <b>Utility functions to zero out our model gradients</b>
+function nullify_grad!(p)
+  if typeof(p) <: TrackedArray
+    p.grad .= 0.0f0
+  end
+  return p
+end
+
+function zero_grad!(model)
```

Comment:
You shouldn't need these utilities; just use `update!` with `Params` and `Grads`, like `Flux.train!` does.

Also, you're still using HTML tags above

Reply:
@MikeInnes I used the update! without zeroing out the gradients of the models. However that model is not converging to the actual output even after repeated trials. However, manually zeroing out the gradients does... I don't know if it's a bug on my part... I was using this as a reference : 
https://github.com/eriklindernoren/PyTorch-GAN/blob/1f130dfca726e14254e4fd78e5fb63f08931acd3/implementations/cgan/cgan.py#L161-L195

As pointed out on Slack, `gradient` used in `update!` should automatically zero out the gradients, but the results are not reflecting them...

Reply:
https://github.com/FluxML/Flux.jl/blob/66ce8d8066e00b0f8eccd69649dabaee26f97129/src/optimise/train.jl#L20 should zero out the gradients with a simple `update!`

Reply:
The normal `update!` method will work if you use `gradient` rather than `back!`. `back!` should be avoided as it's effectively deprecated.

Reply:
@MikeInnes Sorry for replying late.Made the requested changes.

Reply:
I believe zeroing out gradient is necessary here. Denote the parameters for discriminator and generator as `ps_d` and `ps_g` seperately. The `gradient` call on `ps_d` only zeros out the gradient for `ps_d` but that for `ps_g` is not zeroed out. If one then attempts to do more computation and asks for the gradient for `ps_g`, the results is not correct.

Reply:
`update!` does zero out the gradient so no need to do it explicitly I suppose.

Reply:
I believe `update!` only zeros out the gradient in the call but not all the gradient.

Reply:
See this MWE below:

```julia
using Flux, Tracker

d1 = Dense(2, 1)
d2 = Dense(1, 1)
c = Chain(d1, d2)
p1 = params(d1)
p2 = params(d2)
pall = params(c)

x = rand(2, 10)

loss() = sum(c(x))

# Case 1
gradient(loss, pall).grads |> values |> println
# This zeros out all gradients

# Case 2
gradient(loss, p1).grads |> values |> println
# After this call, the gradient of p2 is not zeroed out
# Thus the call for gradient of p2 below will be affected
gradient(loss, p2).grads |> values |> println
# After this call, the gradient of p1 is not zeroed out

# Just zero out all gradients before the next experiment
Tracker.zero_grad!.(Tracker.grad.(p1))
Tracker.zero_grad!.(Tracker.grad.(p2))

# Case 3
gradient(loss, p1).grads |> values |> println
Tracker.zero_grad!.(Tracker.grad.(p2))  # just to avoid the situation in Case 1
gradient(loss, p2).grads |> values |> println
```
gives
```
Any[Float32[10.0] (tracked), Float32[1.8203094] (tracked), Float32[1.2446415 0.7490297] (tracked), Float32[-1.8717546] (tracked)]

Any[Float32[1.8203094] (tracked), Float32[1.2446415 0.7490297] (tracked)]
Any[Float32[20.0] (tracked), Float32[-3.7435093] (tracked)]

Any[Float32[1.8203094] (tracked), Float32[1.2446415 0.7490297] (tracked)]
Any[Float32[10.0] (tracked), Float32[-1.8717546] (tracked)]
```
See how Case 2 is different from Case 1 and Case 3.

Reply:
When I created a DCGAN model with Flux v0.9.0, it didn't converged without zeroing out gradients after training the discriminator. https://github.com/matsueushi/fluxjl-gan/blob/flux0.9.0/mnist-dcgan.jl

However, with Zygote backend (Flux v0.10.0),
```julia
using Flux

d1 = Dense(2, 1)
d2 = Dense(1, 1)
c = Chain(d1, d2)
p1 = params(d1)
p2 = params(d2)
pall = params(c)

x = rand(2, 10)

loss() = sum(c(x))

@info "Case1"
gradient(loss, pall).grads |> values |> println

@info "Case2"
gradient(loss, p1).grads |> values |> println
gradient(loss, p2).grads |> values |> println
```
gives expected results
```shell
[ Info: Case1
Any[Float32[9.715003], Float32[3.7775292 4.41461], Float32[10.0], Float32[-5.8413825]]
[ Info: Case2
Any[Float32[9.715003], Float32[3.7775292 4.41461]]
Any[Float32[10.0], Float32[-5.8413825]]
```
and I didn't need to zero out gradients. https://github.com/matsueushi/fluxjl-gan/blob/e60684b6c8ecc601eb6784ae393eae9a3a3ba57a/mnist-dcgan.jl

Reply:
This is expected. Only a tracker based AD needs the zero-out part. AD based on Zygote doesn't have this side effect.